runtime/parachains/src/inclusion.rs

```diff
@@ -13,3 +13,471 @@
 
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The inclusion module is responsible for inclusion and availability of scheduled parachains
+//! and parathreads.
+//!
+//! It is responsible for carrying candidates from being backable to being backed, and then from backed
+//! to included.
+
+use sp_std::prelude::*;
+use primitives::{
+	parachain::{
+		ValidatorId, AbridgedCandidateReceipt, ValidatorIndex, Id as ParaId,
+		AvailabilityBitfield as AvailabilityBitfield, SignedAvailabilityBitfields, SigningContext,
+		BackedCandidate,
+	},
+};
+use frame_support::{
+	decl_storage, decl_module, decl_error, ensure, dispatch::DispatchResult, IterableStorageMap,
+	weights::Weight,
+	traits::Get,
+};
+use codec::{Encode, Decode};
+use bitvec::{order::Lsb0 as BitOrderLsb0, vec::BitVec};
+use sp_staking::SessionIndex;
+use sp_runtime::{DispatchError, traits::{One, Saturating}};
+
+use crate::{configuration, paras, scheduler::{CoreIndex, GroupIndex, CoreAssignment}};
+
+/// A bitfield signed by a validator indicating that it is keeping its piece of the erasure-coding
+/// for any backed candidates referred to by a `1` bit available.
+#[derive(Encode, Decode)]
+#[cfg_attr(test, derive(Debug))]
+pub struct AvailabilityBitfieldRecord<N> {
+	bitfield: AvailabilityBitfield, // one bit per core.
+	submitted_at: N, // for accounting, as meaning of bits may change over time.
+}
+
+/// A backed candidate pending availability.
+#[derive(Encode, Decode)]
+#[cfg_attr(test, derive(Debug))]
+pub struct CandidatePendingAvailability<H, N> {
+	/// The availability core this is assigned to.
+	core: CoreIndex,
+	/// The candidate receipt itself.
+	receipt: AbridgedCandidateReceipt<H>,
+	/// The received availability votes. One bit per validator.
+	availability_votes: BitVec<BitOrderLsb0, u8>,
+	/// The block number of the relay-parent of the receipt.
+	relay_parent_number: N,
+	/// The block number of the relay-chain block this was backed in.
+	backed_in_number: N,
+}
+
+pub trait Trait: system::Trait + paras::Trait + configuration::Trait { }
+
+decl_storage! {
+	trait Store for Module<T: Trait> as ParaInclusion {
+		/// The latest bitfield for each validator, referred to by their index in the validator set.
+		AvailabilityBitfields: map hasher(twox_64_concat) ValidatorIndex
+			=> Option<AvailabilityBitfieldRecord<T::BlockNumber>>;
+
+		/// Candidates pending availability by `ParaId`.
+		PendingAvailability: map hasher(twox_64_concat) ParaId
```

Comment:
At any time we only permit one candidate per availability core.

Reply:
And only one per para as well.

Parachains each get their own core, so that is synonymous. Parathreads are multiplexed over cores, but the parathread queuing logic in the scheduler ensures that there are no competing cores for parathreads, so it's impossible that a parathread A is scheduled on core X and Y simultaneously.